packages/web3torrent/src/App.tsx

```diff
@@ -33,20 +36,45 @@ const App: React.FC = () => {
       }
     });
   }, [initialized]);
+  const [selectedAddress, setSelectedAddress] = useState(undefined);
+  useEffect(() => {
```

Comment:
Any particular reason to have 2 useEffect hooks that run after every render (as opposed to merging them into 1)?

Reply:
Also, would you mind explaining why we need `useEffect` here? My understanding of what this does is:
1. The App component renders.
1. After the render this hook is run.
1. If `window.ethereum.selectedAddress` has changed, the App component is rerendered.

What about just using the `window.ethereum.selectedAddress` during the first render?

Reply:
Without `useEffect` we run into an infinite render loop.  I think this is due to `setSelectedAddress` being called, which results in the hook below calling `setBalance` which forces a re-render. By using an empty dependency array we limit this hook to running once after the first render.

I don't think we can rely on `window.ethereum.selectedAddress` changing to force a re-render, as I don't think it's a valid dependency for a hook. (It can be mutated, which wouldn't trigger the hook)

So we rely on attaching the listener that updates a piece of state indicating the address has changed. We could change that so it's just a flag for the hook that the address has changed and and just use `window.ethereum.selectedAddress` in the hook but that seems more convoluted that just storing the address and depending on that. That would allow us to avoid the initial hook though.

Reply:
> Any particular reason to have 2 useEffect hooks that run after every render (as opposed to merging them into 1)?

This is mostly to keep the logic organized. One hook is setting up listeners and one hook is setting the current address.

Reply:
Thanks for the explanation, that makes a lot of sense! What about the following code to remove the hook
```
  const [selectedAddressState, setSelectedAddress] = useState(undefined);
  const selectedAddress = selectedAddressState ?? window.ethereum?.selectedAddress;
```

While I do like React hooks better than the class alternative, I think my preference is to use hooks as sparingly as possible since they are not super easy to reason about.